src/ray/object_manager/plasma/object_lifecycle_manager.cc

```diff
@@ -0,0 +1,281 @@
+  }
+  return true;
+}
+
+std::string ObjectLifecycleManager::EvictionPolicyDebugString() const {
+  return eviction_policy_->DebugString();
+}
+
+const LocalObject *ObjectLifecycleManager::CreateObjectInternal(
+    const ray::ObjectInfo &object_info, plasma::flatbuf::ObjectSource source,
+    bool allow_fallback_allocation) {
+  // Try to evict objects until there is enough space.
+  for (int num_tries = 0; num_tries <= 10; num_tries++) {
+    auto result =
+        object_store_->CreateObject(object_info, source, /*fallback_allocate*/ false);
+    if (result != nullptr) {
+      return result;
+    }
+    // Tell the eviction policy how much space we need to create this object.
+    std::vector<ObjectID> objects_to_evict;
+    int64_t space_needed =
+        eviction_policy_->RequireSpace(object_info.GetObjectSize(), &objects_to_evict);
+    EvictObjects(objects_to_evict);
+    // More space is still needed.
+    if (space_needed > 0) {
+      RAY_LOG(DEBUG) << "attempt to allocate " << object_info.GetObjectSize()
+                     << " failed, need " << space_needed;
+      break;
+    }
+    // NOTE(ekl) if we can't achieve this after a number of retries, it's
+    // because memory fragmentation in dlmalloc prevents us from allocating
+    // even if our footprint tracker here still says we have free space.
+  }
+
+  if (!RayConfig::instance().plasma_unlimited()) {
+    RAY_LOG(DEBUG) << "Fallback allocation is not enabled.";
+    return nullptr;
+  }
+  if (!allow_fallback_allocation) {
+    RAY_LOG(DEBUG) << "Fallback allocation not enabled for this request.";
+    return nullptr;
+  }
+
+  RAY_LOG(INFO)
+      << "Shared memory store full, falling back to allocating from filesystem: "
+      << object_info.GetObjectSize();
+
+  auto result =
+      object_store_->CreateObject(object_info, source, /*fallback_allocate*/ true);
+
+  if (result == nullptr) {
+    RAY_LOG(ERROR) << "Plasma fallback allocator failed, likely out of disk space.";
+  }
+  return result;
+}
+
+void ObjectLifecycleManager::EvictObjects(const std::vector<ObjectID> &object_ids) {
+  for (const auto &object_id : object_ids) {
+    RAY_LOG(DEBUG) << "evicting object " << object_id.Hex();
+    auto entry = object_store_->GetObject(object_id);
+    // TODO(rkn): This should probably not fail, but should instead throw an
+    // error. Maybe we should also support deleting objects that have been
+    // created but not sealed.
+    RAY_CHECK(entry != nullptr) << "To evict an object it must be in the object table.";
+    RAY_CHECK(entry->state == ObjectState::PLASMA_SEALED)
+        << "To evict an object it must have been sealed.";
+    RAY_CHECK(entry->ref_count == 0)
+        << "To evict an object, there must be no clients currently using it.";
+
+    DeleteObjectInternal(object_id);
+  }
+}
+
+void ObjectLifecycleManager::DeleteObjectInternal(const ObjectID &object_id) {
+  auto entry = object_store_->GetObject(object_id);
+  if (entry == nullptr) {
```

Comment:
Looks like every caller already guarantees it is nullptr. Consider changing it to RAY_CHECK?